### PR TITLE
Chat text selection

### PR DESCRIPTION
**Pull request by jasper-lyons**
This is rough implementation of text selection in all labels with copy and paste. I still need to:

- [x] make the code meet the coding guidelines
- [x] test the code in all mods
- [x] move multiple classess out of single files
- [x] take on board code / architecture comments

Fixing Highlighting Colurs in `Metrics.yaml`
- [x] RA
- [x] CNC
- [x] D2K
- [x] TS

Issues Identified:
- [x] When the cursor leaves the left most bound of a label without going below the label it should select all of the text from the start point to the beginning of the text.
- [x] Selection should work from the first click on the label.
- [ ] ~When double clicking on a word it shold highlight the word.~
- [ ] ~When tripple clicking on a label it should highlight the whole label.~
- [ ] ~When mousing over a selectable label the mouse cursor should change to a line / cursor indicator.~
- [ ] ~When a label is bigger than the scroll box and some of the label is selected the selection moves with any scrolling movment.~
- [x] Selection should work over multiple lines.
- [x] The highlight position should align with the mouse cursor.

### CODE REVIEW

**Comment by dragunoff** [comment]
This would be a great feature, thanks for taking on the task. You should however make sure not to change the formatting and white-space of the files. Pehaps you text editor has changed the tabs to double-spaces. OpenRA has an `.editorconfig` file so check if your editor has support for that (most modern IDE/editors have some support). This will make sure that white-space remains as it should be and only your code contributions are committed to git.

**Comment by jasper-lyons** [comment]
My bad. Looks like while I was flailing getting `make check` to run I reformatted the whole files without regarding the `.editorconfig`.

That should be fixed now (and this should subsequently be a bit easier to read). Thanks for the heads up!

**Comment by jasper-lyons** [comment]
Found an issue.

I'm handling clicks on labels to facilitate selection but once a click is handled once we're assuming it's done and dusted and not propagating that click up to the parent. This means that the labels handle the click for DropDownButtonWidgets and prevent the triggering of the items that the label is labelling.

Options I see for solving this:

1. Create a specialized SelectableLableWidget which inherits from LabelWidget and encapsulated the selection behaviour.
2. Always return false from label mouse handling functions.
3. Adapt the framework code to allow bubbling to parents even if a child handled an event.
4. Adapt the DropDownButtonWidgets to use MouseUp events for triggering selection since by then we can identify whether the intention was to make a selection or not.

The easy answer for me is (1) but it means we don't get selection everywhere for free and it means a bit more work updating the chat interface to use the new Widget.

In an ideal world I'd say (3) because given a hierarchy of elements in a UI there can be multiple sitations where you want to encapsulate input event behaviour to the respective element instead of handling it in the deepest and manually propagating it up. This isn't free though so *shrug*.

(4) is probably a happy medium between (1) and (3), meaning we get selection in all widgets everywhere for free without breaking DropDownButtonWigets.

(2) is a hack and I would prefer not to do it.

**Comment by pchote** [comment]
(1) sounds best to me. It doesn't make sense to have every label in the UI selectable, and this may cause unexpected problems.

**Comment by jasper-lyons** [comment]
It has the smallest footprint. I'll miss not having map names be selectable etc but I guess other people can update them to use SelectableLabelWidget if they want.

Or maybe I will because I want to... but in a separate PR.

**Comment by jasper-lyons** [comment]
Ok, I've updated everything with your suggestions and along with the coding guidelines.

1 visual issue has arrisen but it's not a "show stopper". In the CnC and TS mods it isn't obvious when we're selecting the sever name as the back ground color acround the text is the same as the highligh color for the text.

I'm not 100% how to go about solving that yet and I think that could be a fixed in a different stream of work. What does anyone else think?

**Comment by dragunoff** [comment]
You can make the highlight color and background color configurable in metrics.yaml

**Comment by jasper-lyons** [comment]
I will have a peek at that though I remember something from the original PR which worked with some configurable values from metrics.yaml

Are there are ~ up to date docs on how to work with metrics.yaml for configuration values? If not, could you point me at some example code? Perhaps a configurable value in a materics.yaml file, the code that actually does the loading and where that metric is used in the code base?

I can find it on my own but it will take me a while so I would appreciate the help some someone who knows what they're doing!

**Comment by dragunoff** [comment]
@jasper-lyons Have a look at  https://github.com/OpenRA/OpenRA/blob/3a6d88cfefa3fff2158d2095e9f89415ec19e9dd/OpenRA.Mods.Common/Widgets/LabelWithHighlightWidget.cs#L23 for an example of how to use the values from chrome metrics.

**Comment by jasper-lyons** [comment]
Fixed for TS but CnC is still causing issues for selecting the ServerName Label.

The problem is:
* The server name is positioned outside of it's parents bounds (Y = -22)
* Click events are only propagated to children if the click event occurs within the parents bounds.

These come to together to mean that the ServerName Label is uninteractable with the mouse.

Possible approaches I've thought of to handle this:
1. Add a root container that can contain the title and the background to enable the ServerName Label to be drawn within it's parents bounds.
2. Adapt the engine to faciliate click events on elements that are children displayed outside of the parents bounds.

I am not sure how to proceed tbh.

**Comment by dragunoff** [comment]
I don't think that the server name should be a blocker for getting this feature in. It's more important to be able to select in the chat. We can track and try to solve the event propagation issue separately. So IMO option 1 (add a wrapper container) or simply ignore the problem for now.

File: OpenRA.Mods.Common/Widgets/LabelWidget.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/LabelWidget.cs]: This doesn't look correct

File: OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs]: Do we have a use case for making these dynamic?

File: mods/ts/mod.yaml
Comment (pchote) [inline mods/ts/mod.yaml]: Why does this need to be cloned?

File: OpenRA.Mods.Common/Widgets/LabelWidget.cs
Comment (jasper-lyons) [reply OpenRA.Mods.Common/Widgets/LabelWidget.cs]: Yup, looks broke. Will fix.

File: OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs
Comment (jasper-lyons) [reply OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs]: You mean the `Func<Type> PropertyName` stuff or the `Type?` stuff?

I copied the `Func<Type>` stuff from the previous PR under the assumption that it was "they way". I'm not aware of a meaningful usecase for them though.

Should I remove?

File: mods/ts/mod.yaml
Comment (jasper-lyons) [reply mods/ts/mod.yaml]: I need to set a property indivudually on the SERVER_NAME Widget of the lobby.yaml. This property is unique to the TS mod.

I couldn't find a way to do this without cloning the whole file, is there an easier way?

**Comment by jasper-lyons** [comment]
So it turns out the issue is that the text / font in labels does not implcitly change the render bounds of a label. All I had to do was give the label an explcit height (and adjust the Y value so that it's position relative to the background didn't change) and the Widget is now visible.

I wonder if not implictly adjusting a Widgets Bounads / RenderBounds based on it's content is something we want on purpose or if that would be useful over all.

**Comment by jasper-lyons** [comment]
Beyond the outstanding comment from @pchote I think this is basically ready for the next steps, maybe someone else pull and test or merge it. I'm not sure of protocol.

**Comment by dragunoff** [comment]
@jasper-lyons I get an NRE when I try to access the lobby (skirmish or multiplayer). Here is the log: 

<details>
```

Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object
  at OpenRA.ObjectCreator.CreateBasic (System.Type type) [0x00000] in /home/ivaylo/code/OpenRA/OpenRA.Game/ObjectCreator.cs:124
  at OpenRA.ObjectCreator.CreateObject[T] (System.String className, System.Collections.Generic.Dictionary`2[TKey,TValue] args) [0x00064] in /home/ivaylo/code/OpenRA/OpenRA.Game/ObjectCreator.cs:101
  at OpenRA.WidgetLoader.NewWidget (System.String widgetType, OpenRA.Widgets.WidgetArgs args) [0x00015] in /home/ivaylo/code/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:86
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, OpenRA.MiniYamlNode node) [0x00026] in /home/ivaylo/code/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:53
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, OpenRA.MiniYamlNode node) [0x00128] in /home/ivaylo/code/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:70
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, System.String w) [0x0002a] in /home/ivaylo/code/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:45
  at OpenRA.Widgets.Ui.OpenWindow (System.String id, OpenRA.Widgets.WidgetArgs args) [0x00000] in /home/ivaylo/code/OpenRA/OpenRA.Game/Widgets/Widget.cs:62
  at OpenRA.Game.OpenWindow (System.String widget, OpenRA.Widgets.WidgetArgs args) [0x00000] in /home/ivaylo/code/OpenRA/OpenRA.Game/Game.cs:127
  at OpenRA.Mods.Common.Widgets.Logic.MainMenuLogic.OpenSkirmishLobbyPanel () [0x00007] in /home/ivaylo/code/OpenRA/OpenRA.Mods.Common/Widgets/Logic/MainMenuLogic.cs:452
  at OpenRA.Mods.Common.Widgets.Logic.ConnectionLogic.ConnectionStateChanged (OpenRA.Network.OrderManager om) [0x00014] in /home/ivaylo/code/OpenRA/OpenRA.Mods.Common/Widgets/Logic/ConnectionLogic.cs:28
  at (wrapper delegate-invoke) System.Action`1[OpenRA.Network.OrderManager].invoke_void_T(OpenRA.Network.OrderManager)
  at OpenRA.Game.LogicTick () [0x0002f] in /home/ivaylo/code/OpenRA/OpenRA.Game/Game.cs:623
  at OpenRA.Game.Loop () [0x000f1] in /home/ivaylo/code/OpenRA/OpenRA.Game/Game.cs:791
  at OpenRA.Game.Run () [0x0003c] in /home/ivaylo/code/OpenRA/OpenRA.Game/Game.cs:832
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in /home/ivaylo/code/OpenRA/OpenRA.Game/Game.cs:262
  at OpenRA.Program.Main (System.String[] args) [0x00014] in /home/ivaylo/code/OpenRA/OpenRA.Game/Support/Program.cs:31
[ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object
  at OpenRA.ObjectCreator.CreateBasic (System.Type type) [0x00000] in /home/ivaylo/code/OpenRA/OpenRA.Game/ObjectCreator.cs:124
  at OpenRA.ObjectCreator.CreateObject[T] (System.String className, System.Collections.Generic.Dictionary`2[TKey,TValue] args) [0x00064] in /home/ivaylo/code/OpenRA/OpenRA.Game/ObjectCreator.cs:101
  at OpenRA.WidgetLoader.NewWidget (System.String widgetType, OpenRA.Widgets.WidgetArgs args) [0x00015] in /home/ivaylo/code/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:86
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, OpenRA.MiniYamlNode node) [0x00026] in /home/ivaylo/code/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:53
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, OpenRA.MiniYamlNode node) [0x00128] in /home/ivaylo/code/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:70
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, System.String w) [0x0002a] in /home/ivaylo/code/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:45
  at OpenRA.Widgets.Ui.OpenWindow (System.String id, OpenRA.Widgets.WidgetArgs args) [0x00000] in /home/ivaylo/code/OpenRA/OpenRA.Game/Widgets/Widget.cs:62
  at OpenRA.Game.OpenWindow (System.String widget, OpenRA.Widgets.WidgetArgs args) [0x00000] in /home/ivaylo/code/OpenRA/OpenRA.Game/Game.cs:127
  at OpenRA.Mods.Common.Widgets.Logic.MainMenuLogic.OpenSkirmishLobbyPanel () [0x00007] in /home/ivaylo/code/OpenRA/OpenRA.Mods.Common/Widgets/Logic/MainMenuLogic.cs:452
  at OpenRA.Mods.Common.Widgets.Logic.ConnectionLogic.ConnectionStateChanged (OpenRA.Network.OrderManager om) [0x00014] in /home/ivaylo/code/OpenRA/OpenRA.Mods.Common/Widgets/Logic/ConnectionLogic.cs:28
  at (wrapper delegate-invoke) System.Action`1[OpenRA.Network.OrderManager].invoke_void_T(OpenRA.Network.OrderManager)
  at OpenRA.Game.LogicTick () [0x0002f] in /home/ivaylo/code/OpenRA/OpenRA.Game/Game.cs:623
  at OpenRA.Game.Loop () [0x000f1] in /home/ivaylo/code/OpenRA/OpenRA.Game/Game.cs:791
  at OpenRA.Game.Run () [0x0003c] in /home/ivaylo/code/OpenRA/OpenRA.Game/Game.cs:832
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in /home/ivaylo/code/OpenRA/OpenRA.Game/Game.cs:262
  at OpenRA.Program.Main (System.String[] args) [0x00014] in /home/ivaylo/code/OpenRA/OpenRA.Game/Support/Program.cs:31
AL lib: ReleaseALC: 1 device not closed

 ```
</details>

**Comment by dragunoff** [comment]
Also please run `make check` and make sure it passes before pushing code to github. Otherwise the travis build fails and the PR can't be merged.

**Comment by jasper-lyons** [comment]
Disappointing.

Rebasing to remove some stuff I re-introduced some stlye cop issues. My bad, fixed now.

In removing the dynamic properties I seem to have caused an issue that you experienced above. I've added the dynamic properties in again but I honestly don't understand why they are required.

**Comment by jasper-lyons** [comment]
It appears the issue was removing the constructor for the type which makes sense, can't create an instance of a contructor if there isn't a contructor for you to get an instance of.

File: mods/ts/mod.yaml
Comment (dragunoff) [reply mods/ts/mod.yaml]: This highlights a limitation of the chrome metrics - they set a single value for a property but often times we may need more than one. Just like here where a different highlight color is needed depending on the background color. It is also a limitation of chrome yaml - some kind of yaml inheritance for defining components or default styles/properties would be useful here.

**Comment by jasper-lyons** [comment]
Finally added a pre-push hook to run `make check` which should avoid failed builds in the future.

I've fixed the bugs and issues that were present in the previous implementation though there is a weird quirk in multiline selection where the calculation from the curcor to the nearest line when outside the bounds of the label picks lines other than the nearest line. I have yet to diagnose this issue but it seems to be infrequent for me.

Testing of that would be greatly appreciated.

File: mods/ts/mod.yaml
Comment (jasper-lyons) [reply mods/ts/mod.yaml]: I was thinking about this and it feel a little like what CSS was invented to solve alongside HTML, applying properties selectively to elements in a tree without having to copy the whole (or parts of) the tree.

**Comment by jasper-lyons** [comment]
Ok. This is done I think. I'm going to move the remaining specs for this into new issue once this is merged / on the roadmap.

File: OpenRA.Game/Graphics/SpriteFont.cs
Comment (dragunoff) [inline OpenRA.Game/Graphics/SpriteFont.cs]: Please revert this white-space change. Even though it does look somewhat better :man_shrugging: Also I see a lot of double-tab indentations elsewhere in your code. I'm not sure what are our conventions for multi-line formatting :thinking: Someone with an opinion should step up here.

File: OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs]: Unnecessary using directive.

File: OpenRA.Game/Widgets/WidgetLoader.cs
Comment (dragunoff) [inline OpenRA.Game/Widgets/WidgetLoader.cs]: Unnecessary using directive.

File: OpenRA.Mods.Common/Widgets/Selection.cs
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/Selection.cs]: Unnecessary using directive.

File: OpenRA.Mods.Common/Widgets/LabelWidget.cs
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/LabelWidget.cs]: Do we need to address anything here or has this already been figured out? If there is a todo the comment should be prefixed with a `TODO: `.

File: OpenRA.Mods.Common/Widgets/LabelWidget.cs
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/LabelWidget.cs]: Same about todos here.

File: OpenRA.Mods.Common/Widgets/LabelWidget.cs
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/LabelWidget.cs]: Same about todos here.

File: OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs]: This can be a `readonly` prop.

File: mods/ts/metrics.yaml
Comment (dragunoff) [inline mods/ts/metrics.yaml]: We can go for a color that works on all backgrounds to avoid duplicating `lobby.yaml`. Something around `#777777` seems to be OK.

**Review by dragunoff — CHANGES_REQUESTED** [review]
I tested on Windows and Linux and it works quite smoothly. Though the selection still lags behind the cursor on Linux the further to the right I drag. 

I added a few comments, mostly nits, I'll let someone more experienced to review the logic parts.

File: OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs:26]: Could you rename these to `TextSelectionColor` and `TextSelectionBackgroundColor`? I think it makes the purpose clearer and is more consistent with existing names such as `TextColor` and `TextHighlightColor`.

File: OpenRA.Game/Graphics/SpriteFont.cs
Comment (tovl) [reply OpenRA.Game/Graphics/SpriteFont.cs]: Opinionated person chiming in :smile: 
As far as I'm aware, this formatting change is correct. I use tab on line continuation all the time and never gotten flak for it. I think it is a good convention to follow.

Double-tab is definitely gratuitous, but this may be just an editor issue regarding 4-space tabs vs 8-space tabs and not a deliberate thing.

**Comment by jasper-lyons** [comment]
Thats annoying. Haha. I'm deving on a linux system so I'd be interested to explore why the cursor is still lagging on it!

Otherwise, thanks for the review. I'm going to go through and address things that I can immediately.

File: OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs
Comment (jasper-lyons) [reply OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs]: Probably to allow for using `Console.Write`. Will fix.

File: OpenRA.Mods.Common/Widgets/Selection.cs
Comment (jasper-lyons) [reply OpenRA.Mods.Common/Widgets/Selection.cs]: Left over from pulling out the Selection.cs code from LabelWithSelection.cs. Will fix.

File: OpenRA.Mods.Common/Widgets/LabelWidget.cs
Comment (jasper-lyons) [reply OpenRA.Mods.Common/Widgets/LabelWidget.cs]: These were actually left over from the original PR that I inherited to implement this. I'm unsure of their purpose, they can probably be removed tbh unless someone who understands what's going on better understands what they are referencing.

File: OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs
Comment (jasper-lyons) [reply OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs]: Makes sense.

File: mods/ts/metrics.yaml
Comment (jasper-lyons) [reply mods/ts/metrics.yaml]: Would we want to do the same for the `Textfield` since they have selection functionality aswell and we'd want matching backgrounds, no?

I'm assuming this is so that changes to the lobby can be made across all mods without having to change many files?

File: OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs
Comment (jasper-lyons) [reply OpenRA.Mods.Common/Widgets/LabelWithSelectionWidget.cs:26]: Had the same thoughts just didn't get around to it. Will fix.

File: mods/ts/metrics.yaml
Comment (dragunoff) [reply mods/ts/metrics.yaml]: Yes, the same color should be applied to the textfield selection. 

TS, RA and D2k share many of the chrome layout files (found in the common folder). TD is the odd one out as it implements a different UI (that is probably for historical reasons but it also serves as an example of how to make an alternative UI). Duplicating the chrome yamls (with such a small change) is a maintenance burden.

File: OpenRA.Game/Graphics/SpriteFont.cs
Comment (pchote) [inline OpenRA.Game/Graphics/SpriteFont.cs:261]: There are two issues with doing this character by character:
* It's really inefficient
* If the text flows over multiple lines the selection background for lower lines hides the tails of long letters like `j`.

I think it would be better to avoid adding this new method (at least until we have more use cases), and instead have `LabelWithSelectionWidget` measure the width of the selected text in each line and draw a single rectangle for each line (and all lines) before using the normal DrawText on top.

File: OpenRA.Mods.Common/Widgets/Selection.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Selection.cs:139]: I'd prefer it if we could define `var bottomRight = topLeft + font.Measure(val.ToString())` and then use `var bounds = Rectangle.FromLTRB(topLeft.X, topLeft.Y, bottomRight.X, bottomRight.Y)`. Adding lots of similar-but-different ways to create objects can cause real headaches when refactoring code, and I don't think the 3 uses in this file really justify adding a new method.

File: OpenRA.Mods.Common/Widgets/Selection.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Selection.cs:131]: We will need to answer this before we can merge: I suspect the answer will be "break". I don't yet have a feeling for how this is used, but in general assuming that "text character" == `char` is a bad move when working with modern text.

File: mods/cnc/chrome/lobby.yaml
Comment (pchote) [inline mods/cnc/chrome/lobby.yaml:16]: I'd prefer it if we didn't do this. I think #17684 should fix the underlying issue properly?

**Comment by pchote** [comment]
I'm sad to see progress stall here after all the effort that went into it, but as its now been over half a year without progress i'll close this to help clean up the PR queue 🙁 

@teinarss has picked this up and has been working lately on the remaining issues. They can hopefully file a new PR once things are in a good state for merging.

**Pull request closed without merging**
